### PR TITLE
➕ Add IconAlignment to new Material Buttons

### LINKED ISSUE #71941: Allow icon button constructor to specify a left or right icon
Currently we can not create Buttons with an icon at the right, well we can if we build it by our own, but flutter already have one constructor to help with that, the missing part is that it only allows you to put icons at the left of the button.

I was thinking to add the code and create a PR so I wanted to get some feedback before.

# Current
`FlatButton` or others have a named constructor `FlatButton.icon` which lets you specify an icon.
This icon is always on the left, the code looks like:

https://github.com/flutter/flutter/blob/8f3a12f1447b4baad06910eadd48b30a0a95e305/packages/flutter/lib/src/material/flat_button.dart#L300-L307

# Expected
Have one way to show it in the left or in the right, possible options:
- Add two named constructors, `leftIcon` and `rightIcon`.
- Add two parameters, `leftIcon` and `rightIcon` and only one of those can be supplied.

### PR DESCRIPTION
**Pull request by YazeedAlKhalaf**
## Description

I have added a new property to three material buttons:
* `TextButton`
* `ElevatedButton`
* `OutlinedButton`

The new property allows the user to align the icon to the left or to the right 🚀, which appears to be useful in some usages.

## Related Issues

Fixes #71941

## Tests

I added the following tests:

- `TextButton.icon is configurable when iconAlignment set to ButtonIconAlignment.start`
- `TextButton.icon is configurable when iconAlignment set to ButtonIconAlignment.end`
- `OutlinedButton.icon is configurable when iconAlignment set to ButtonIconAlignment.start`
- `OutlinedButton.icon is configurable when iconAlignment set to ButtonIconAlignment.end`
- `ElevatedButton.icon is configurable when iconAlignment set to ButtonIconAlignment.start`
- `ElevatedButton.icon is configurable when iconAlignment set to ButtonIconAlignment.end`

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/material/outlined_button.dart [hunk 1]
```diff
@@ -7,6 +7,7 @@ import 'dart:ui' show lerpDouble;
 
 import 'package:flutter/foundation.dart';
 import 'package:flutter/rendering.dart';
+import 'package:flutter/src/material/button.dart';
```
Comment (RoyARG02) [inline packages/flutter/lib/src/material/outlined_button.dart:10]: > For old buttons(FlatButton, RaisedButton, and OutlineButton) in flutter/packages/flutter/lib/src/material/button.dart. 
> For new buttons(TextButton, ElevatedButton, and OutlinedButton) in `flutter/packages/flutter/lib/src/material/button_style_button.dart`.

As per my comment on the issue, this shouldn't be included here. You were probably trying to import this in **OutlineButton**.

I'm kind of conflicted, because doing as quoted above would result in code duplication.
Someone having a better experience with material buttons may be able to guide here.

File: packages/flutter/lib/src/material/text_button.dart [hunk 2]
```diff
@@ -7,6 +7,7 @@ import 'dart:ui' show lerpDouble;
 
 import 'package:flutter/foundation.dart';
 import 'package:flutter/rendering.dart';
+import 'package:flutter/src/material/button.dart';
```
Comment (RoyARG02) [inline packages/flutter/lib/src/material/text_button.dart:10]: Ditto of my previous comment.

Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/text_button.dart:10]: I have done this in the three files, where do i need to put the enum in your opinion?

File: packages/flutter/lib/src/material/button.dart [hunk 3]
```diff
@@ -589,3 +589,19 @@ class _RenderInputPadding extends RenderShiftedBox {
     );
   }
 }
+
+/// How the icon in any `button.icon` should be placed where `button` could be:
```
Comment (RoyARG02) [inline packages/flutter/lib/src/material/button.dart:593]: Probably a better fit at `button_style_button.dart`, if it is included _only_ for the new buttons.

File: packages/flutter/lib/src/material/text_button.dart (hunk 2 above)
Comment (RoyARG02) [reply packages/flutter/lib/src/material/text_button.dart:10]: Sorry, the review UI didn't update and I missed your reply. You could take a look at my comment below.

File: packages/flutter/lib/src/material/outlined_button.dart (hunk 1 above)
Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/outlined_button.dart:10]: I believe the enum is in the best place right now

File: packages/flutter/lib/src/material/button.dart (hunk 3 above)
Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/button.dart]: I took a look at the file and it seems that it is a stateful widget and not really related. I think the `button.dart` file is the best place

**Comment by YazeedAlKhalaf** [comment]
Should I add this functionality to old material buttons or leave them so people get encouraged to use new material buttons?

File: packages/flutter/lib/src/material/button.dart (hunk 3 above)
Comment (RoyARG02) [reply packages/flutter/lib/src/material/button.dart]: Sorry, but I couldn't follow your point.
The following lines show that _it_ clearly relates to the new buttons:
https://github.com/flutter/flutter/blob/a12e2a473a3214b4556ca589d2d9bd311ac7c6f1/packages/flutter/lib/src/material/button_style_button.dart#L21-L30
The new buttons inherit this class. For instance
https://github.com/flutter/flutter/blob/a12e2a473a3214b4556ca589d2d9bd311ac7c6f1/packages/flutter/lib/src/material/text_button.dart#L57

Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/button.dart]: so you suggest I move the enum to `button_style_button.dart` from `button.dart`?

Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/button.dart]: CC @Hixie

Comment (RoyARG02) [reply packages/flutter/lib/src/material/button.dart]: As long as this feature is meant for the new buttons only, I think you should.
However, it may or may not be useful if it is decided to bring this feature to old buttons as well.

File: packages/flutter/lib/src/material/button.dart [hunk 4]
```diff
@@ -589,3 +589,19 @@ class _RenderInputPadding extends RenderShiftedBox {
...
+
+/// How the icon in any `button.icon` should be placed, where `button` could be:
+/// * [TextButton]
+/// * [ElevatedButton]
+/// * [OutlinedButton]
+///
+/// See also:
+///
+/// * [TextButton], [ElevatedButton], [OutlinedButton], the button widgets
+enum IconAlignment {
```
Comment (RoyARG02) [inline packages/flutter/lib/src/material/button.dart:601]: Can I recommend naming this along the lines of something like `ButtonIconAlignment`, so as to make sense to the user in where can this enum be applied to?
Also, owing the possibility of text directionality(ltr/rtl), using `ButtonIconAlignment.left` and `ButtonIconAlignment.right` will probably be misleading(`ButtonIconAlignment.left` would display icon on the left in ltr, but right in rtl).
Something like `ButtonIconAlignment.leading/trailing` might be a better name.

Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/button.dart]: Yea, thanks a lot. It makes much sense. Will do the changes but will use `start` and `end` to make like the `MainAxisAlignment` and `CrossAxisAlignment` since people are used to those.

File: packages/flutter/lib/src/material/button.dart (hunk 3 above)
Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/button.dart]: I found the right file for the `enum`, it is` button_style.dart`. I will be adding it there as a final change for its location where it fits the most. As you said the `button.dart` file is for the obsolete material buttons.

**Comment by YazeedAlKhalaf** [comment]
@HansMuller can you review this PR?

**Comment by Hixie** [comment]
@YazeedAlKhalaf This will definitely need tests. Any new functionality needs tests, because otherwise we might accidentally remove the functionality without noticing. That's what the tests are for: to make sure accidental regressions don't happen.

File: packages/flutter/lib/src/material/text_button.dart [hunk 5]
```diff
@@ -377,18 +379,30 @@ class _TextButtonWithIconChild extends StatelessWidget {
...
+
+    List<Widget> _renderButtonChildren() {
+      switch (iconAlignment) {
+        case ButtonIconAlignment.start:
+          return <Widget>[icon, SizedBox(width: gap), label];
+        case ButtonIconAlignment.end:
+          return <Widget>[label, SizedBox(width: gap), icon];
+      }
+    }
+
```
Comment (Hixie) [inline packages/flutter/lib/src/material/text_button.dart:402]: drive-by nit: this would probably be more performant if you set a variable rather than using a function here.

Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/text_button.dart:402]: Sure, thanks. Will do the change 🚀

**Comment by YazeedAlKhalaf** [comment]
@Hixie I added some tests for all the new buttons that should make removing the additions I did break the code so I guess that is what the tests are for! 🚀

File: packages/flutter/lib/src/material/text_button.dart (hunk 5 above)
Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/text_button.dart:402]: I am trying to find a more performant way but can't, can you reply with an example if that is possible 😊🙌

File: packages/flutter/lib/src/material/button_style.dart [hunk 6]
```diff
@@ -435,3 +435,19 @@ class _LerpShapes implements MaterialStateProperty<OutlinedBorder?> {
...
 }
+
+/// How the icon in any `button.icon` should be placed, where `button` could be:
+/// * [TextButton]
+/// * [ElevatedButton]
+/// * [OutlinedButton]
+///
+/// See also:
+///
+/// * [TextButton], [ElevatedButton], [OutlinedButton], the button widgets
```
Comment (Hixie) [inline packages/flutter/lib/src/material/button_style.dart:446]: we probably don't need the See also section here since it just repeats the list just above.

File: packages/flutter/lib/src/material/button_style.dart [hunk 7]
```diff
@@ -435,3 +435,19 @@ class _LerpShapes implements MaterialStateProperty<OutlinedBorder?> {
     return ShapeBorder.lerp(resolvedA, resolvedB, t) as OutlinedBorder?;
   }
 }
+
+/// How the icon in any `button.icon` should be placed, where `button` could be:
+/// * [TextButton]
```
Comment (Hixie) [inline packages/flutter/lib/src/material/button_style.dart:440]: you need a blank line before a bulleted list in our markdown syntax.

File: packages/flutter/lib/src/material/button_style.dart [hunk 8]
```diff
@@ -435,3 +435,19 @@ class _LerpShapes implements MaterialStateProperty<OutlinedBorder?> {
...
+/// * [ElevatedButton]
+/// * [OutlinedButton]
+///
+/// See also:
+///
+/// * [TextButton], [ElevatedButton], [OutlinedButton], the button widgets
+enum ButtonIconAlignment {
+  /// Places the icon to the start of the button
```
Comment (Hixie) [inline packages/flutter/lib/src/material/button_style.dart:448]: please end all sentences with a period.

File: packages/flutter/lib/src/material/button_style.dart [hunk 9]
```diff
@@ -435,3 +435,19 @@ class _LerpShapes implements MaterialStateProperty<OutlinedBorder?> {
...
+  start,
+
+  /// Places the icon to the end of the button
+  end,
+}
```
Comment (Hixie) [inline packages/flutter/lib/src/material/button_style.dart:453]: the documentation here for "start" and "end" doesn't say anything you couldn't guess from the name. This violates our style guide. :-) https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-useless-documentation

Try to think of why someone would be looking at this documentation, and answer questions they may have. For example, what does "end of the button" mean? Is there padding applied? Is it inside the button or outside the button?

File: packages/flutter/lib/src/material/text_button.dart [hunk 10]
```diff
@@ -377,18 +379,30 @@ class _TextButtonWithIconChild extends StatelessWidget {
...
     return Row(
       mainAxisSize: MainAxisSize.min,
-      children: <Widget>[icon, SizedBox(width: gap), label],
+      children: _renderButtonChildren(),
     );
   }
```
Comment (Hixie) [inline packages/flutter/lib/src/material/text_button.dart:407]:
```dart
  @override
  Widget build(BuildContext context) {
    final double scale = MediaQuery.maybeOf(context)?.textScaleFactor ?? 1;
    final double gap = scale <= 1 ? 8 : lerpDouble(8, 4, math.min(scale - 1, 1))!;
    List<Widget> children;
    switch (iconAlignment) {
      case ButtonIconAlignment.start:
        children = <Widget>[icon, SizedBox(width: gap), label];
        break;
      case ButtonIconAlignment.end:
        children = <Widget>[label, SizedBox(width: gap), icon];
        break;
    }
    return Row(
      mainAxisSize: MainAxisSize.min,
      children: children,
    );
  }
```

File: packages/flutter/test/material/text_button_test.dart [hunk 11]
```diff
@@ -1009,6 +1009,46 @@ void main() {
...
+            label: const Text('text button'),
+          ),
+        ),
+      ),
+    );
+    final dynamic textButtonWithIconWidget = tester.widget(find.byWidgetPredicate((Widget widget) => '${widget.runtimeType}' == '_TextButtonWithIconChild'));
+    expect(textButtonWithIconWidget.iconAlignment, buttonIconAlignment);
+  });
+
 }
```
Comment (Hixie) [inline packages/flutter/test/material/text_button_test.dart:1053]: your tests verify that the value is passed into the _TextButtonWithIconChild, but not that the _TextButtonWithIconChild itself is doing anything with it, which is really the most important thing to check. :-)

File: packages/flutter/lib/src/material/button_style.dart (hunk 6 above)
Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/button_style.dart:446]: Okay, will remove it.

File: packages/flutter/lib/src/material/button_style.dart (hunk 7 above)
Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/button_style.dart:440]: Sure, I will add it 🚀.

File: packages/flutter/lib/src/material/button_style.dart (hunk 8 above)
Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/button_style.dart:448]: Sure I will 🚀

File: packages/flutter/lib/src/material/text_button.dart (hunk 10 above)
Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/text_button.dart:407]: Okay, thanks 🚀. Changes Added.

File: packages/flutter/test/material/text_button_test.dart (hunk 11 above)
Comment (YazeedAlKhalaf) [reply packages/flutter/test/material/text_button_test.dart:1053]: Ah okay I get it, I will try to check that the items place changes:
* when `iconAlignment` is set to `ButtonIconAlignment.start` it should be [Icon, SizedBox, Text].
* when `iconAlignment` is set to `ButtonIconAlignment.end` it should be [Text, SizedBox, Icon].
I think this will be good, will add it and let you know 🚀.

Comment (YazeedAlKhalaf) [reply packages/flutter/test/material/text_button_test.dart:1079]: Great news 🎉. I finally got some time to do the test and I think they work great! I hope you can take a look.

File: packages/flutter/lib/src/material/button_style.dart (hunk 9 above)
Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/button_style.dart:467]: I guess the new docs I just added are good, let me know what you think.

**Comment by YazeedAlKhalaf** [comment]
I added all required changes, hopefully, the checks are all successful. I will request re-request review from @Hexie and hope that @HansMuller can take a look. Thanks, everybody 🚀.

File: packages/flutter/lib/src/material/button_style.dart [hunk 12]
```diff
@@ -435,3 +435,33 @@ class _LerpShapes implements MaterialStateProperty<OutlinedBorder?> {
     return ShapeBorder.lerp(resolvedA, resolvedB, t) as OutlinedBorder?;
   }
 }
+
+/// How the icon in any `button.icon` should be placed, where `button` could be:
```
Comment (Hixie) [inline packages/flutter/lib/src/material/button_style.dart:439]: the first line of the docs is what will appear in the the API docs summary screen, so we should make that a self-contained sentence, like "Position of the icon in a button relative to the label".

File: packages/flutter/lib/src/material/button_style.dart [hunk 13]
```diff
@@ -435,3 +435,33 @@ class _LerpShapes implements MaterialStateProperty<OutlinedBorder?> {
...
+///
+/// * [TextButton]
+/// * [ElevatedButton]
+/// * [OutlinedButton]
```
Comment (Hixie) [inline packages/flutter/lib/src/material/button_style.dart:443]: It could be used by other buttons too, maybe say something like `This is used by [TextButton], [ElevatedButton], and [OutlinedButton].` so that it's not implied to be exclusively those?

File: packages/flutter/lib/src/material/button_style.dart [hunk 14]
```diff
@@ -435,3 +435,33 @@ class _LerpShapes implements MaterialStateProperty<OutlinedBorder?> {
...
+///
+/// In buttons that have an icon, we generally tend to use a `Row` widget
```
Comment (Hixie) [inline packages/flutter/lib/src/material/button_style.dart:445]: avoid "we" (we use the passive voice in the API docs).

But also, I would generally avoid mentioning how this is implemented, because we might change it.

File: packages/flutter/lib/src/material/button_style.dart [hunk 15]
```diff
@@ -435,3 +435,33 @@ class _LerpShapes implements MaterialStateProperty<OutlinedBorder?> {
...
+/// to be able to put multiple widgets in a horizontal way, aka `Row`.
+///
+/// Hover over enum values to be able to see more deailed description
+/// on what do they do.
```
Comment (Hixie) [inline packages/flutter/lib/src/material/button_style.dart:449]: Avoid saying how to use the docs -- "hover" doesn't work on api.flutter.dev for example.

File: packages/flutter/lib/src/material/button_style.dart [hunk 16]
```diff
@@ -435,3 +435,33 @@ class _LerpShapes implements MaterialStateProperty<OutlinedBorder?> {
...
+enum ButtonIconAlignment {
+  /// Places the icon to the `start` of the button which means the
+  /// first child in the `Row`.
+  ///
+  /// The `Row` will be like this if you choose to use `start` which is
```
Comment (Hixie) [inline packages/flutter/lib/src/material/button_style.dart:454]: avoid "you" in the API docs if possible.

File: packages/flutter/lib/src/material/button_style.dart [hunk 17]
```diff
@@ -435,3 +435,33 @@ class _LerpShapes implements MaterialStateProperty<OutlinedBorder?> {
...
+  /// the default:
+  ///
+  /// `[YOUR_ICON, SIZEDBOX, LABEL]`
```
Comment (Hixie) [inline packages/flutter/lib/src/material/button_style.dart:457]: This is nice and detailed but I think we'd be better off avoiding saying exactly how it's implemented. Another type of button could use a different implementation but this enum could still apply.

Also we should say what "start" means.

File: packages/flutter/lib/src/material/button_style.dart [hunk 18]
```diff
@@ -435,3 +435,33 @@ class _LerpShapes implements MaterialStateProperty<OutlinedBorder?> {
...
+  start,
+
+  /// Places the icon to the `end` of the button which means the
+  /// last child in the `Row`.
+  ///
+  /// The `Row` will be like this if you choose to use `end`:
+  ///
+  /// `[LABEL, SIZEDBOX, YOUR_ICON]`
+  end,
```
Comment (Hixie) [inline packages/flutter/lib/src/material/button_style.dart:466]: similar comments here

File: packages/flutter/lib/src/material/elevated_button.dart [hunk 19]
```diff
@@ -393,7 +395,7 @@ class _ElevatedButtonWithIcon extends ElevatedButton {
          focusNode: focusNode,
          autofocus: autofocus ?? false,
          clipBehavior: clipBehavior ?? Clip.none,
-         child: _ElevatedButtonWithIconChild(icon: icon, label: label),
+         child: _ElevatedButtonWithIconChild(icon: icon, label: label, iconAlignment: iconAlignment ?? ButtonIconAlignment.start),
```
Comment (Hixie) [inline packages/flutter/lib/src/material/elevated_button.dart:398]: can the default just be in the constructor?

File: packages/flutter/lib/src/material/outlined_button.dart [hunk 20]
```diff
@@ -325,7 +327,7 @@ class _OutlinedButtonWithIcon extends OutlinedButton {
          focusNode: focusNode,
          autofocus: autofocus ?? false,
          clipBehavior: clipBehavior ?? Clip.none,
-         child: _OutlinedButtonWithIconChild(icon: icon, label: label),
+         child: _OutlinedButtonWithIconChild(icon: icon, label: label, iconAlignment: iconAlignment ?? ButtonIconAlignment.start),
```
Comment (Hixie) [inline packages/flutter/lib/src/material/outlined_button.dart:330]: same comment

File: packages/flutter/lib/src/material/text_button.dart [hunk 21]
```diff
@@ -345,18 +346,19 @@ class _TextButtonWithIcon extends TextButton {
...
+        assert(label != null),
+        super(
+          key: key,
+          onPressed: onPressed,
+          onLongPress: onLongPress,
+          style: style,
+          focusNode: focusNode,
+          autofocus: autofocus ?? false,
+          clipBehavior: clipBehavior ?? Clip.none,
+          child: _TextButtonWithIconChild(icon: icon, label: label, iconAlignment: iconAlignment ?? ButtonIconAlignment.start),
```
Comment (Hixie) [inline packages/flutter/lib/src/material/text_button.dart:360]: same comment

File: packages/flutter/lib/src/material/text_button.dart [hunk 22]
```diff
@@ -377,18 +379,31 @@ class _TextButtonWithIconChild extends StatelessWidget {
...
+    List<Widget> children;
+
+    switch (iconAlignment) {
+      case ButtonIconAlignment.start:
+        children = <Widget>[icon, SizedBox(width: gap), label];
+        break;
+      case ButtonIconAlignment.end:
+        children = <Widget>[label, SizedBox(width: gap), icon];
+        break;
+    }
```
Comment (Hixie) [inline packages/flutter/lib/src/material/text_button.dart:402]: is there any way we can avoid duplicating this code everywhere?

File: packages/flutter/test/material/text_button_test.dart [hunk 23]
```diff
@@ -1009,6 +1009,72 @@ void main() {
...
+      find.descendant(
+        of: textButtonWithIconFinder,
+        matching: find.byType(Row),
+      ),
+    );
+    expect(label, rowInsideTextButton.children.first);
+    expect(icon, rowInsideTextButton.children.last);
+  });
+
 }
```
Comment (Hixie) [inline packages/flutter/test/material/text_button_test.dart:1078]: we should also have a test that verifies that the alignment works when you change the text direction

Comment (YazeedAlKhalaf) [reply packages/flutter/test/material/text_button_test.dart:1078]: text direction will flip it

File: packages/flutter/lib/src/material/elevated_button.dart (hunk 19 above)
Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/elevated_button.dart:398]: i have it as required so to be able to accomplish what is requested I have to make the value optional, do I proceed?

File: packages/flutter/lib/src/material/button_style.dart (hunk 15 above)
Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/button_style.dart:449]: Yea it makes sense 🤣

File: packages/flutter/lib/src/material/button_style.dart (hunk 14 above)
Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/button_style.dart:445]: I believe it is useful right now but I could add a todo or a note that states changing the docs of ButtonIconAlignment to be without implementation if used with other types of buttons that do not use `Row`.

So what do you think?

Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/button_style.dart:445]: I have changes as requested, now without implementation

File: packages/flutter/lib/src/material/button_style.dart (hunk 16 above)
Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/button_style.dart:454]: thanks, I have removed the whole block because as requested it has an implementation that could change and be used by other types of buttons

File: packages/flutter/lib/src/material/button_style.dart (hunk 17 above)
Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/button_style.dart:457]: thanks, I have updated the docs to state what `start` means by talking about the horizontal axis and where relative to the button it will be.

File: packages/flutter/lib/src/material/button_style.dart (hunk 18 above)
Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/button_style.dart:466]: done 🚀

File: packages/flutter/lib/src/material/text_button.dart (hunk 22 above)
Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/text_button.dart:402]: I don't know but from what I can see, buttons have a lot of duplications. I think it is okay because otherwise, we need to use a method that returns something. We could put it in `button_style.dart`.

File: packages/flutter/test/material/text_button_test.dart (hunk 23 above)
Comment (Hixie) [reply packages/flutter/test/material/text_button_test.dart:1078]: For example, check the position of the icon with one direction, then check it with the opposite direction.

File: packages/flutter/lib/src/material/text_button.dart (hunk 22 above)
Comment (Hixie) [reply packages/flutter/lib/src/material/text_button.dart:402]: Using a method is a good way to reduce code duplication.

File: packages/flutter/lib/src/material/elevated_button.dart (hunk 19 above)
Comment (Hixie) [reply packages/flutter/lib/src/material/elevated_button.dart:398]: If it's required then you don't need to use `??`, right? What is the API you intend to expose?

File: packages/flutter/test/material/text_button_test.dart (hunk 23 above)
Comment (YazeedAlKhalaf) [reply packages/flutter/test/material/text_button_test.dart:1078]: ah okay

File: packages/flutter/lib/src/material/elevated_button.dart (hunk 19 above)
Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/elevated_button.dart:398]: I said it wrong, I mean required in `_ElevatedButtonWithIconChild` but not in `_ElevatedButtonWithIcon` or `ElevatedButton.icon` which makes it mandatory somewhere to check for `null` using `??`.

File: packages/flutter/lib/src/material/text_button.dart (hunk 22 above)
Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/text_button.dart:402]: ok sure, 🙈

Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/text_button.dart]: Done 🚀

File: packages/flutter/test/material/text_button_test.dart (hunk 23 above)
Comment (YazeedAlKhalaf) [reply packages/flutter/test/material/text_button_test.dart:1078]: I mean the `textDirection` flips the concept if you look at it from a `ltr` language, but for example, if you look at it when text direction is 'rtl' it makes sense that start is at the right because that's where people start reading from in `rtl` languages. I think it is not needed but it is your call ✌. The point of using `ButtonIconAlignment` is to put the icon at the end or at the start complying with the text direction. I really don't get it why we need a test here? 😉🙌

File: packages/flutter/lib/src/material/text_button.dart (hunk 22 above)
Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/text_button.dart]: I have also made changes to the template because yesterday a pull request has been merged changing only `OutlinedButton.icon`
#70872

**Comment by HansMuller** [comment]
To reverse the default order in which the icon and label widgets appear, why not just write

```dart
TextButton.icon(label: bar, icon: foo)
```
instead of 
```dart
TextButton.icon(label: foo, icon: bar)
```

This PR seems like a rather lot of API and implementation to get the same effect?

**Comment by YazeedAlKhalaf** [comment]
> To reverse the default order in which the icon and label widgets appear, why not just write
> 
> 
> 
> ```dart
> 
> TextButton.icon(label: bar, icon: foo)
> 
> ```
> 
> instead of 
> 
> ```dart
> 
> TextButton.icon(label: foo, icon: bar)
> 
> ```
> 
> 
> 
> This PR seems like a rather lot of API and implementation to get the same effect?

I don't see the difference in code blocks above

**Comment by YazeedAlKhalaf** [comment]
> To reverse the default order in which the icon and label widgets appear, why not just write
> 
> ```dart
> TextButton.icon(label: bar, icon: foo)
> ```
> 
> instead of
> 
> ```dart
> TextButton.icon(label: foo, icon: bar)
> ```
> 
> This PR seems like a rather lot of API and implementation to get the same effect?

It works but does not make sense. This new API would allow for new buttons to use it. If we create a button where the label and icon are not widgets but rather a `String` and an `IconData` type, this will make it more usable and makes a lot more sense to use in that way.

File: packages/flutter/lib/src/material/elevated_button.dart (hunk 19 above)
Comment (YazeedAlKhalaf) [reply packages/flutter/lib/src/material/elevated_button.dart:402]: What do you suggest doing?

**Comment by Hixie** [comment]
@YazeedAlKhalaf So now that we have a patch to show how much code this takes, we have to consider the trade-offs of resolving the issue vs leaving it as-is (without this feature). (We have to do this for all features.)

* Pro: It allows one to switch the direction of the icons and buttons without using TextDirection or misusing the two fields on the button.
* Con: It adds a bunch of code (289 lines across multiple files), including a new enum (top-level in the API docs, which is especially expensive for when people try to find something in the docs). More code means less performance and bigger binary sizes, as well as the opportunity for more bugs, slower velocity when adding new features, and more complexity for people using Flutter.
* Con: It doesn't seem to be a feature that the Material specs recommend. There's no examples of having the icons on the "end" side of a button, for instance. There's one with buttons above the text rather than to the left of the text, but that is marked as a thing to avoid doing.

Given that it's relatively easy to make buttons do this by just passing them a label that's a Row, or by swapping the icon and label parameters (a weird thing to do but it does work), or by using Directionality, I think overall the value of this change is probably not enough to warrant the cost. I understand that that is painful to hear (I've experienced this myself many times, where I'll spend a bunch of time on a branch building something only to decide that it's just not worth it :-( ).

If you agree, you can close the PR. That said, if there's something I missed, e.g. some important reason why this PR does offset its cost, please do let me know! For example, if icons and labels are more complicated to align than just swapping them around, or if this is a highly-thumbed-up issue.

**Comment by Piinks** [comment]
Hi @YazeedAlKhalaf, this has not been updated in a while, and is not in a state for code review, so I am going to close it for now. If you would like to come back and continue discussing this change, feel free to re-open! Thank you for the contribution!

**Comment by PassionPenguin** [comment]
it's an important feature... simply adding a flag in constructor would help a lot... rather than change the widgets between `icon` and `label` (which is not good for semantics)

**Comment by YazeedAlKhalaf** [comment]
> it's an important feature... simply adding a flag in constructor would help a lot... rather than change the widgets between `icon` and `label` (which is not good for semantics)

@Hixie

**Comment by PassionPenguin** [comment]
> it's an important feature... simply adding a flag in constructor would help a lot... rather than change the widgets between `icon` and `label` (which is not good for semantics)

it could be fixed simply by changing `icon` & `widget` name in textbutton class to `widget1` and `widget2` and assign to these two final variable by system with the constraints of `textdirection` and `iconalignment`....

but it could make a big diff, eg, in pagination (`< Prev` `1` `2` `3` `4` `5` `Next >`)

**Comment by YazeedAlKhalaf** [comment]
The extra flag would also allow for more functionality such as putting the icon above or under the text.

**Comment by Kitllers** [comment]
yezzirrr
